File: frontend/web/monorepo/libs/mateu/src/mateu/shared/apiClients/dtos/componentmetadata/ResponsiveGrid.ts
/**
 * One responsive grid — THE general layout foundation (coherence-plan #9). The tracks (sized with
 * the #8 hug/fixed/fill vocabulary) arrive resolved as a CSS grid-template-columns string; children
 * travel as the component's children.
 */
export default interface ResponsiveGrid {
    gridTemplateColumns?: string | undefined
    gap?: string | undefined
    /** Optional per-child column span, aligned with the children; a child with span N occupies N
     *  tracks (a full-width band uses a large span). */
    colSpans?: number[] | undefined
    /** Responsive breakpoint: a CSS length below which the grid (its container) collapses to one
     *  column. Absent = never collapse. */
    stackBelow?: string | undefined
    /** Named-slot template (coherence-plan #7): a CSS grid-template-areas value; a child whose slot
     *  matches an area name is placed there. Absent = children flow into the column tracks. */
    gridTemplateAreas?: string | undefined
    /** Named areas pinned with position:sticky while the rest of the grid scrolls (coherence-plan
     *  #7): a child whose slot is listed here gets a sticky wrapper. Absent/empty = none. */
    stickyAreas?: string[] | undefined
}

File: backend/shared/uidl/src/main/java/io/mateu/uidl/data/ResponsiveGrid.java
package io.mateu.uidl.data;

import io.mateu.uidl.fluent.Component;
import java.util.List;

/**
 * One responsive grid — THE general layout foundation (coherence-plan #9). Children are placed on a
 * CSS grid whose column tracks are sized with the #8 vocabulary ({@link GridTrack}: hug=auto,
 * fill=1fr, fixed=len), so form columns, zones, dashboards and template slots can all be expressed
 * as one grid instead of five specialized mechanisms. This is the foundation + escape hatch, not
 * the default authoring mode — inference still builds columns by default.
 *
 * <p>v1 carries explicit column tracks and an optional per-child column span ({@code colSpans},
 * aligned with {@code content}: a child with span N occupies N tracks — this is what a full-width
 * band or a wide dashboard tile needs, so the scattered column mechanisms can converge here).
 * Responsive breakpoints and named areas are follow-ups.
 */
public record ResponsiveGrid(
    String id,
    List<GridTrack> columns,
    String gap,
    List<Component> content,
    List<Integer> colSpans,
    String style)
    implements Component {

  public ResponsiveGrid(String id, List<GridTrack> columns, List<Component> content) {
    this(id, columns, null, content, null, null);
  }

  public ResponsiveGrid(
      String id, List<GridTrack> columns, String gap, List<Component> content, String style) {
    this(id, columns, gap, content, null, style);
  }

  /**
   * The CSS {@code grid-template-columns} value resolved from the tracks (e.g. "auto 1fr 15rem").
   */
  public String gridTemplateColumns() {
    if (columns == null || columns.isEmpty()) {
      return null;
    }
    return columns.stream().map(GridTrack::toCss).reduce((a, b) -> a + " " + b).orElse(null);
  }
}
